data-processing/common/upload_entities.py

```diff
@@ -155,73 +109,193 @@ def upload_entities(
+) -> EntityModels:
+    """
+    Build a map from the within-paper entity IDs to database row IDs. It is assumed that the
+    number of entities already uploaded for this paper won't be so many that they can't all
+    be stored in memory.
+    """
+
+    if data_version is None:
+        data_version = get_or_create_data_version(s2_id)
+
+    entity_models: EntityModels = {}
     uploaded_entities = Entity.select().where(
         Entity.paper_id == s2_id, Entity.version == data_version
     )
     for entity_model in uploaded_entities:
-        uploaded_entity_models[
-            (entity_model.type, entity_model.within_paper_id)
-        ] = entity_model
+        entity_models[(entity_model.type, entity_model.within_paper_id)] = entity_model
+
+    return entity_models
+
+
+def make_data_models(
+    entity_model: Optional[Entity], entity_model_id: Optional[str], data: EntityData
+) -> List[EntityDataModel]:
+    " Either 'entity_model' or 'entity_model_id' must be defined. "
+
+    if entity_model is None and entity_model_id is None:
+        logging.warning(  # pylint: disable=logging-not-lazy
+            "Attempted to create data models without providing an entity model or explicit ID. "
+            + "At least one of these must be provided to determine which entity the data should "
+            + "be associated with."
+        )
+        return []
+
+    models: List[EntityDataModel] = []
 
-    def resolve_model(entity_type: str, within_paper_id: str) -> Optional[Entity]:
+    # Create models for each field in the entity data.
+    for key, value in data.items():
+
+        # The value may have been specified as a list, or as a single scalar value.
+        # Unpack all of the values for this key into a list.
+        values: List[Any] = []
+        if isinstance(value, list):
+            of_list = True
+            values.extend(value)
+        else:
+            of_list = False
+            values.append(value)
+
+        if len(values) == 0:
+            continue
+
+        value_types = {type(v) for v in values}
+        if not len(value_types) == 1:
+            logging.warning(  # pylint: disable=logging-not-lazy
+                "Attempted to upload multiple primitive types of data for key %s. "
+                + "Types were %s. Not permitted. Skipping this value.",
+                key,
+                value_types,
+            )
+            continue
+
+        # Create a new row for each value, with information of the base data type
+        # and whether that row belongs to a list. If casting of values needs to occur
+        # for values based on type to make them appropriate for insertion in Postgres
+        # (e.g., casting booleans to 0 / 1), that should happen here.
+        for v in values:
+            type_ = None
+            # Check for boolean needs to come before check for int, because booleans
+            # will pass the check 'isinstance(v, int)'.
+            if isinstance(v, bool):
```

Comment:
This all seems weird to me, but I don't understand enough about the end-to-end or API expected data format to know why it's like this! Why do we need item type?

Reply:
Good question! It's a bit convoluted:

At one point in time, I thought we might want to add arbitrary new data fields of any primitive type for entities, and then have them served up to the API without having to make changes to the API code.

Therefore, the API currently has code that scans through all entitydata rows for each entity, then uses the type listed in the `item_type` column to cast that item to the right type (e.g., a boolean, an integer, a string) before packaging that data up into a JSON response to send to the UI.

This design was to make it faster to prototype the UI as we discovered we needed to expose more information about the entities. I'm not sure it makes sense in the long term for the API, as I imagine we would the API to only expose certain data fields for each entity, rather than any type of data found in the entitydata table. Does that all make sense?